fontbe/src/features.rs

```diff
@@ -111,22 +302,167 @@ fn write_debug_fea(context: &Context, is_error: bool, why: &str, fea_content: &s
+    let kerned_locations: HashSet<_> = kerning.kerns.values().flat_map(|v| v.keys()).collect();
+    let mut kerned_locations: Vec<_> = kerned_locations.into_iter().collect();
+    kerned_locations.sort();
+
+    if log::log_enabled!(log::Level::Trace) {
+        trace!(
+            "The following {} locations have kerning:",
+            kerned_locations.len()
+        );
+        for pos in kerned_locations.iter() {
+            trace!("  {pos:?}");
+        }
+    }
+
+    // For any kern that is incompletely specified fill in the missing values using UFO kerning lookup
+    // Not 100% sure if this is correct for .glyphs but lets start there
+    // Generate variable format kerning per https://github.com/adobe-type-tools/afdko/pull/1350
+    // Use design values per discussion on https://github.com/harfbuzz/boring-expansion-spec/issues/94
+    let mut fea = String::new();
+    fea.reserve(8192); // TODO is this a good value?
+    fea.push_str("\n\n# fontc generated kerning\n\n");
+
+    if kerning.is_empty() {
+        return Ok(fea);
+    }
+
+    // TODO eliminate singleton groups, e.g. @public.kern1.Je-cy = [Je-cy];
+
+    // 1) Generate classes (http://adobe-type-tools.github.io/afdko/OpenTypeFeatureFileSpecification.html#2.g.ii)
+    // @classname = [glyph1 glyph2 glyph3];
+    for (name, members) in kerning.groups.iter() {
+        fea.push_str("@");
+        fea.push_str(name.as_str());
+        fea.push_str(" = [");
+        for member in members {
+            fea.push_str(member.as_str());
+            fea.push_str(" ");
+        }
+        fea.remove(fea.len() - 1);
+        fea.push_str("];\n");
+    }
+    fea.push_str("\n\n");
+
+    // 2) Generate pairpos (http://adobe-type-tools.github.io/afdko/OpenTypeFeatureFileSpecification.html#6.b)
+    // it's likely that many kerns use the same location string, might as well remember the string edition
+    let mut pos_strings = HashMap::new();
+    fea.push_str("feature kern {\n");
+    for ((participant1, participant2), values) in kerning.kerns.iter() {
+        fea.push_str("  pos ");
+        push_identifier(&mut fea, participant1);
+        fea.push_str(" ");
+        push_identifier(&mut fea, participant2);
+
+        // See https://github.com/adobe-type-tools/afdko/pull/1350#issuecomment-845219109 for syntax
+        // <value>n for normalized, per https://github.com/harfbuzz/boring-expansion-spec/issues/94#issuecomment-1608007111
+        fea.push_str(" (");
+        for location in kerned_locations.iter() {
+            let pos_str = pos_strings.entry(location).or_insert_with(|| {
+                location
+                    .iter()
+                    .map(|(tag, value)| format!("{tag}={}n", value.into_inner()))
+                    .collect::<Vec<_>>()
+                    .join(",")
+            });
+
+            // TODO can we skip some values by dropping where value == interpolated value?
```

Comment:
you'd have to construct a new temporary model with a given master omitted, interpolate at that location and compare original with interpolated value, might become expensive to do for many kerns * masters. We don't currently do any of that in the python pipeline, source masters are always retained even if potentially redundant.

Reply:
If we let people hand-write variable value records I think we *have* to support building new variation models in case they write a rule that references a point that isn't aligned with any master. To be fair, we could kick that out for now as I don't believe current compilation or sources allow this.